sycl/doc/design/PropagateCompilerFlagsToRuntime.md

```diff
@@ -0,0 +1,124 @@
+# Propagation of optimization levels used by front-end compiler to backend
+
+In order to ease the process of debugging, there is a user requirement to
+compile different modules with different levels of optimization. This document
+proposes a compiler flow that will enable propagation of compiler options
+specified for front-end to the runtimes and eventually to the backend.
+Currently, only `O0`/`O1`/`O2`/`O3` options are handled.
+
+## Background
+
+When building an application with several source and object files, it is
+possible to specify the optimization parameters individually for each source
+file/object file (for each invocation of the DPCPP compiler). The SYCL runtime
+should pass the original optimization options (e.g. `-O0` or `-O2`) used when
+building an object file to the device backend compiler. This will improve the
+debugging experience by selectively disabling/enabling optimizations for each
+source file, and therefore achieving better debuggability and better performance
+as needed.
+
+The current behavior is that the optimization level option is captured at link
+time and converted into its backend-specific equivalent. This option is
+propagated to the backend. For example, If `-O0` option is specified during
+link-time when using the OpenCL backend, the SYCL runtime will pass
+`-cl-opt-disable` option to the backend device compiler for all modules
+essentially disabling optimizations globally. Otherwise, if the `-O0`
+option is not specified for linker, it will not pass `-cl-opt-disable` option at
+all, therefore making the kernels mostly undebuggable, regardless of the
+original front-end compiler options. Link-time capturing of optimization option
+is the essence of the current implementation and this leads to loss of 
+information about the compile-time options. Proposed design aims to rectify this
+behavior.
+
+Here is an example that demonstrates this pain point:
+
+```
+clang++ -c test_host.cpp -o test_host.o
+clang++ -c -fsycl test_device_1.cpp -o test_device_1.o
+clang++ -c -fsycl -g -O0 test_device_2.cpp -o test_device_2.o
+clang++ -fsycl -g -o test_host.o test_device_1.o test_device_2.o -o test
```

Comment:
```suggestion
clang++ -fsycl -g test_host.o test_device_1.o test_device_2.o -o test
```

I think?

Reply:
it takes multiple object files and links them and returns `test`. Thanks